README.md
# WORD ADDICTS

## App Platform: Android

## Link to GitHub of WordAddicts Project : [GitHub Pages] (https://github.com/TAMMYLE/WordAddicts.git)

## *Project Overview*

You want to have something interesting, yet a little brain twisting for the morning you wake up your brain or you need something in your lunch break to refresh from morning shift, this game get you covered. You will discover the beauty of words, have fun with letters and immerse in the world of vocabulary.

- Student 1: Phong Le - 216324041 
- Student 2: Tammy Le - 216404524

## *LetterStorage class*

* Function randomWord() *that choose a Random word from an array of words*.

     * Parameters: no parameters needed for this function

     * Returns:
     return a String contains pseudorandom word
     
```js
public static String randomWord(){
        return WORDS[RANDOM.nextInt(WORDS.length)];
    }
```

* Function shuffleWord() *function that shuffles the chosen word*

     * Parameters: word - a string storing a word

     * Returns:
     if word is not empty and the length is not space => converting word to a new character array
     then, shuffling the word
     
```js
public static String shuffleWord(String word) {

        if (word != null && !"".equals(word)){
        
            char a[] = word.toCharArray();

            for (int i = 0; i < a.length; i++){
                int j = RANDOM.nextInt(a.length);
                char tmp = a[i];
                a[i]= a[j];
                a[j] = tmp;
            }

            return new String(a);
        }

        return word;
    }
```
* Function compareWords() *function compare two given words*

     *  Parameters: two words that need to be compared

     * Returns: a boolean value indicating if two given words are match

```js
public static Boolean compareWords(String word1, String word2){
        word1 = word1.toLowerCase();
        word2 = word2.toLowerCase();
        char a[] = word1.toCharArray();
        char b[] = word2.toCharArray();

        boolean match = true;
        if(a.length != b.length || a.length == 0 || b.length == 0   )
        {
            match = false;
            return match;
        }

        for (int i = 0; i < a.length; i++)
        {
            if(a[i] != b[i])
                match = false;
            return match;

        }
        return match;
    }
```

* Function giveHint() *review a letter each time the function is called*

     * Parameters : 

          - **givenWord** is what the right word is
          - **hintField** is what already revealed
          - **coin** that the player has

     * Return : void

```js
public void giveHint(String givenWord, TextView hintField, int coin)
    {
        //if there is not enough coins, there will be no hint given
        if(coin <= 0)
        {
            return;
        }
        // get the string from the hint field
        String hintRevealed = hintField.getText().toString(); 
        
        // turn the givenWord into an array of chars
        char[] a = givenWord.toCharArray(); 
        
        // turn what on the hint field into an array of char
        char[] b = hintRevealed.toCharArray(); 
        
        // a new array which has the length of (b+1)
        char[] c = new char[b.length +1 ]; 



        //if given word is longer than what hints already revealed than another letter
        //assign all elements of a to c.length to c
        if(a.length > b.length)
        {
            for (int i = 0; i<c.length; i++ )
            {
                c[i] = a[i];
            }
        }

        //take array c as new hint with one added letter
        hintRevealed = String.copyValueOf(c);
        hintField.setText(hintRevealed);
    }
```

## *Main class*

* Function renewWord() *is to renew the word that player has to guess*

    * parameters: no parameters needed
    
    * return: void
    
```js
public void renewWord() {
        givenWord = availableWords.randomWord();
        shuffedWord = availableWords.shuffleWord(givenWord);
        shuffedLetters.setText(shuffedWord);
    }
```

* Event check is triggered when player clicks check button
     * *if the answer is correct --> score plus ten*

     * *set the Score field with extra score*

     * *clear the input*

     * *renew the given word*

     * *the if condition is triggered when user guess the right word (guessWord == givenWord)*

```js
checkButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                guessWord = textInput.getText().toString();
                //display a message according to the result
                if(availableWords.compareWords(givenWord, guessWord))
                {                                       
                    score += 10;                        //if the answer is correct, score plus ten
                    result.setText("Score: " + score);  //set the Score field with extra score
                    textInput.setText("");              //clear the input
                    renewWord();                        
                }
                else
                {
                    result.setText("Score" + score);

                    //When player gets a wrong word -> input will be vibrated
                   
                    Animation vibrate = AnimationUtils.loadAnimation(getApplicationContext(), R.anim.vibrate);
                    textInput.startAnimation(vibrate);
                }
            }
        });
```
* Method editCoin() *take the current amount of coin and put it into the SharedPreferences*

     * Parameters: current number of coin in Integer form

     * Return: void, the function just edit the coin in the SharedPreference, doesn't return anything

```js
public void editCoin(int coin)
    {
        SharedPreferences preferences = getSharedPreferences("COIN_PREFS", 0);
        SharedPreferences.Editor editor = preferences.edit();
        editor.putInt("coin", coin);
        editor.apply();
    }
```
* Method retrieveCoin() *retrieve the amount of coin currently holding in the SharedPreference*

     * Parameters: no parameters needed

     * Return: integer contains the number of coins.

```js
public int retrieveCoin()
    {
        int coin1;
        SharedPreferences preferences = getSharedPreferences("COIN_PREFS", 0);

        coin1 = preferences.getInt("coin", 0);
        return coin1;
    }
```

## *Timer class*

* Function setTimer: is used to initialize values for the timer
     * parameters: no parameters needed for this funtion
     *  returns : void
     *  const: time --> set the time to 10 which will be multiplied by 1000 to make the timer to 10 seconds
 
 ```js
     private void setTimer(){

        //initialize the time variable
        int time = 10;

        //calculate the totaltime
        totalTimeCountInMilliseconds =  time * 1000;

        //set the maximum time of the progress bar timer
        mProgressBar1.setMax( time * 1000);
    }
```
* Function startTimer: is used to start counting down the timer
     * parameters: no parameters needed for this function
     * returns : void
 
 ```js
 private void startTimer() {

        //call renewWord function once the timer is on
        renewWord();
        countDownTimer = new MyCountDown(totalTimeCountInMilliseconds, 50);
        countDownTimer.start();
    }
```

* Function plusTime: is used to give extra time whenever player gets correct word
     *   parameters: no parameters needed for this function
     *   returns : void

```js
private void plusTime(){

        if(totalTimeCountInMilliseconds < 10000)
        {
            countDownTimer.cancel();
            countDownTimer = new MyCountDown(totalTimeCountInMilliseconds + 3000, 50);
        }
        else
        {
            countDownTimer.cancel();
            countDownTimer = new MyCountDown(10000, 50);
        }

        countDownTimer.start();


    }
```

* Start Timer Event: initialize the click listeners for start button
     * sound is played
     * setTimer() function is called
     * set the visibility of all the needed views
     * start the timer by calling startTimer()

```js
buttonStartTime.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                //play sound
                newgame.start();
                
                //call setTimer function
                setTimer();
                
                score = 0;
                scoreTextView.setText("Score: " + score);
                leveltextview.setText("I");
                
                //set the visibility of all needed views
                buttonStartTime.setVisibility(View.INVISIBLE);
                mProgressBar.setVisibility(View.INVISIBLE);
                textViewShuffle.setVisibility(View.VISIBLE);//show the word that user needs to guess
                scoreTextView.setVisibility(View.VISIBLE);// show the score when the game starts
                leveltextview.setVisibility(View.VISIBLE);// show the level when the game starts
                timerCoin.setVisibility(View.VISIBLE);//show the coin when the game starts
                buttonBuyTime.setVisibility(View.VISIBLE);// show the buyTime button when the game starts
                textViewInput.setVisibility(View.VISIBLE);//show the textInput

                //call startTimer function
                startTimer();

                //set the visibility of the main progress bar
                mProgressBar1.setVisibility(View.VISIBLE);



            }
        });
```

* Key Enter Event: 
     * set the onKey listener for textViewInput, 
     * it gets the user input from the keyboard and then compare with the givenWord
     * The level is also displayed according to what score user is having
     * sound is played accordingly
     
```js

textViewInput.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if ((event.getAction() == KeyEvent.ACTION_DOWN) && (keyCode == KeyEvent.KEYCODE_ENTER))
                {
                    //retrieving user's input --> convert to string
                    guessWord = textViewInput.getText().toString();

                    //added by Tammy, 3/5/2019
                    //Hide the soft keyboard when user pressed
                    closeKeyboard();

                    //clear the input
                    textViewInput.setText("");
                    //display a message according to the result
                    if(availableWords.compareWords(givenWord, guessWord))
                    {

                        if(score >= 800)
                        {
                            //set max length for input Level 4
                            textViewInput.setFilters(new InputFilter[]{ new InputFilter.LengthFilter(6) });
                            leveltextview.setText("IV");
                            renewWordLevel4();
                        }

                        else if(score >= 400)
                        {
                            //set max length for input Level 3
                            textViewInput.setFilters(new InputFilter[]{ new InputFilter.LengthFilter(5) });
                            leveltextview.setText("III");
                            renewWordLevel3();
                        }

                        else if(score >= 200)
                        {
                            //set max length for input Level 2
                            textViewInput.setFilters(new InputFilter[]{ new InputFilter.LengthFilter(4) });
                            leveltextview.setText("II" );
                            renewWordLevel2();
                        }
                        else
                        {
                            leveltextview.setText("I");
                            renewWord();
                        }

                        score += 10;//add the score
                        scoreTextView.setText("Score: " + score);//increase the score when user insert a right answer
                        plusTime();
                        //play sound
                        correctWord.start();

                    }
                    else
                    {
                        //play sound incorrect
                        incorrectWord.start();
                    }

                }
                return false;
            }
        });
```
In order to run the app with CardView and Paypal you need to go to **Gradle (Module: App)**. Add these two lines:



```js
    implementation 'com.android.support:cardview-v7:28.0.0'
    implementation 'com.paypal.sdk:paypal-android-sdk:2.16.0'
```
                 

     





CHANGELOG.md

# Week 5, Thursday 4/April

[TAMMY]
- Created Logo and Icon for WordAddicts
- Create Splashsrceen (SplashAcivity) with Logo 
- Splashscreen jumps to Mainscreen (Main Activity)



# Week 6, Monday 8/April
[TAMMY]
- Created LetterStorage class
- Created an array that stores all words
- Created randomWord function
- Created shuffleWord function

[PHONG]

- Created main menu screen
- App icon is added
- compareWords function is added


# Week 6, Tuesday 9/April
[PHONG]
- basic function for simple gameplay is added


# Week 6, Wednesday 10/April
[HENRY]
- Who worked today? I can see code was changed, but I don't see an update here to explain it. At the end of EVERY work day/work session, update this changelog to tell your partner what you changed, so that they know (and so I know). Thanks :)
- Leave my comments in here every week.

[TAMMY]
- I think there was misunderstanding because we have updated the progress below. As you created new "Wednesday 10/April", I will copy and paste them here :)
- Updated Menu xml file: changed the main layout, added ScrollView to the Menu layout
- Link from SplashScreen to Menu, from Menu to Main 

[PHONG]
- reset function is added

# Week 7, Monday 15 April
- Feedback: Please ensure you keep the dates in this file in order. Otherwise you'll get confused. [Henry]
- Feedback: Some functions still missing function comments. [Henry]
- Feedback: I feel your project is falling a bit behind, and you're not putting in enough time to pass. Show me 10-14 hours each per week.. where was it spent? 
- Feedback: I want you to start writing the time taken per changelog item (in brackets at the end of each line), so that you can see where time is being spent / not being spent enough. [Henry]

# Week 7, Monday 15 April
[TAMMY]
- renew function is added: this function is worked when player enters correct word then the given word is reset as well as the the input is cleared. [1.5 hours]
- added Score field which will add extra point for the score everytime player has a correct word. [1.5 hrs]
- Updated the reset function with the score reset to 0. [15 mins]
- added Highscore activity which will show the best 3 highest scores of the game (java/xml) . [2 hrs]
- added Highscore button on MainActivity and used SharedPreferences to save and retrieve the scores. [15 mins]

# Week 7, Tuesday 16 April
[PHONG]
- updated the main function, make it more object-oriented by using function from LettersStorage class [1 hour]
- fixed errors of checkButton functions in Main Activity [15 minutes]
- Hint Function added [3 hours]
- Coin which is used to reveal hints is added

# Week 7, Wednesday 17 April
[Tammy]
- fixed errors in LettersStorage [1 hrs]
- added Timer activity which is an upgrade level for the word game, including using progress circular bar [12 hours]
- Timer activity will include setTimer function, startTimer function, stopTimer (will be considered)
- added extra 2 drawable resources for those circular progress bars [2 hours]
# Week 7, Sunday 21 April
[Phong]
- demo for TypeShift activity is created (unfinished) [6 hours]

[Tammy]
- Fixed an error: clear the input in Timer activity. Before: when time is up and user cannot get the right answer -> input was not cleared by itself. After: when time is up and user cannot get the right answer -> input is cleared  [30 mins]
- Unfinished plusTime function in Timer activity

# Week 8, Tuesday 30 April
[Tammy]
- Added plusTime function in Timer Activity [3 Hrs] 
# Week 8, Wednesday 1 May
[PHONG] 
- Fix the progress bar [2 hours]
# Week 8, Friday 3 May
[Tammy]
- BeginActivity is initialized with video in background - video needs to be fixed[2 hours]
- Updated checkButton with hiding soft keyboard function [30 mins]

[PHONG]
-update more levels on letterStorage class [2 hours]
# Week 8, Sat 4 May
[PHONG]
- update one more level on letterStorage class [0.5 hour]
- fix the playing error on beginActivity class [1 hour]

# Week 9, Mon 6 May

[PHONG]

- added coins and buyTime function for TimerActivity[3 hours]
- added TimerHighScore activity to store high score when user plays the Speed Mode [2 hours]
# Week 9, Tues 7 May

[PHONG]
- design menuActivity[2 hours]
- design the shop activity [2 hours]

# Week 9, Wednesday 8/May [Henry]
- Good progress. Glad to see it's getting closer. Final stretch now. 
- Make sure you go over everything in the Grading Criteria super carefully, to ensure you meet each point.


# Week 9, Friday 10 May
[Tammy]
- Added vibration for wrong words in TimerActivity
- Created AboutActivity which tells information about the project and us
- Added some basic sounds to some buttons
- Re-design the ShopActivity

# Week 10, Monday 13 May
[PHONG]

- Design main activity and timer activity

[Tammy]
- Updated design of progress bar in Timer
- Activity Added Level4 in LetterStorage

# Week 10, Tuesday 14 May
[PHONG]
- Make some changes to the timer activity [1 hour]
- Add SharedPreference for coin [1 hour]

[Tammy]
- Updated LetterStorage
- Added sounds when user get wrong or right answers
- Deleted checkButton in TimerActivity
- Added Enter to check the input

# Week 10, Wednesday 15 May

[Tammy]
- Updated License file
- Added information in Readme file

# Week 10, Thursday 16 May
[PHONG]
- finalise the app
- added paymentActivity


app/src/main/java/com/example/wordaddicts/TimerActivity.java
/*****  Created by Tammy Le, 16/4/2019        *******/
package com.example.wordaddicts;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.MediaPlayer;
import android.os.Build;
import android.os.CountDownTimer;
import android.support.annotation.RequiresApi;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.InputFilter;
import android.text.Spanned;
import android.view.KeyEvent;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

public class TimerActivity extends AppCompatActivity{


    ProgressBar mProgressBar, mProgressBar1;
    int score;
    private int coin;

    private ImageView highScoreTimer;

    private Button buttonStartTime;
    private Button buttonBuyTime;

    private EditText textViewInput;
    private TextView textViewShuffle, scoreTextView, timerCoin;
    private TextView leveltextview;

    private LinearLayout timerShop;

    private String guessWord;          //what user put in
    private String givenWord;          // what word the game wants user to guess
    private String shuffedWord;        // the word the game gives but already shuffled

    private LettersStorage availableWords = new LettersStorage();

    private MyCountDown countDownTimer;

    //declare the total time of the timer counting in milliseconds
    private long totalTimeCountInMilliseconds;




    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_timer);

        //set media player
        final MediaPlayer correctWord = MediaPlayer.create(this, R.raw.correct);
        final MediaPlayer incorrectWord = MediaPlayer.create(this, R.raw.incorrect);
        final MediaPlayer newgame = MediaPlayer.create(this, R.raw.myclicksound);

        //initialize views

        buttonStartTime = (Button) findViewById(R.id.button_timerview_start);
        buttonBuyTime = (Button) findViewById(R.id.buyTimeButton);
        buttonBuyTime.setVisibility(View.INVISIBLE);// hide the button when the game has not started

        mProgressBar = (ProgressBar) findViewById(R.id.progressbar_timerview);
        mProgressBar1 = (ProgressBar) findViewById(R.id.progressbar1_timerview);

        highScoreTimer = (ImageView) findViewById(R.id.highscoreTimer);

        //declare level variable
        //added by Tammy Le, 13/5/2019
        leveltextview = (TextView) findViewById(R.id.leveltextview);
        leveltextview.setText("LEVEL 1");//initialize the level
        leveltextview.setVisibility(View.INVISIBLE);// set the LevelTextView to be invisible until the game starts

        textViewInput = (EditText)
                findViewById(R.id.textViewInput);
        textViewInput.setVisibility(View.INVISIBLE);

        textViewShuffle = (TextView)
                findViewById(R.id.shuffledWord)    ;
        textViewShuffle.setVisibility(View.INVISIBLE);//hiding the word that user needs to guess

        scoreTextView = (TextView) findViewById(R.id.scoreTextView); // take the scoreTextView from the layout
        scoreTextView.setText("Score : 0");// set the initial score to 0;
        scoreTextView.setVisibility(View.INVISIBLE);// set the scoreTextView to be invisible until the game starts
        score = 0;//set the score to 0;

        timerCoin = (TextView) findViewById(R.id.timerCoin);
        coin = retrieveCoin();//retrieve coin from the shared preference
        timerCoin.setText("" + coin);

        timerShop = (LinearLayout) findViewById(R.id.timerShop);

        //Set max length for first level input
        textViewInput.setFilters(new InputFilter[]{ new InputFilter.LengthFilter(4) });

        /*
        set the onKey listener for textViewInput
        it gets the user input from the keyboard and then compare with the givenWord
        The level is also displayed according to what score user is having
        sound is played accordingly
         */
        textViewInput.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if ((event.getAction() == KeyEvent.ACTION_DOWN) && (keyCode == KeyEvent.KEYCODE_ENTER))
                {
                    //retrieving user's input --> convert to string
                    guessWord = textViewInput.getText().toString();

                    //added by Tammy, 3/5/2019
                    //Hide the soft keyboard when user pressed
                    closeKeyboard();

                    //clear the input
                    textViewInput.setText("");
                    //display a message according to the result
                    if(availableWords.compareWords(givenWord, guessWord))
                    {

                        if(score >=800)
                        {
                            //set max length for input Level 4
                            textViewInput.setFilters(new InputFilter[]{ new InputFilter.LengthFilter(6) });
                            leveltextview.setText("IV");
                            renewWordLevel4();
                        }

                        else if(score >=400)
                        {
                            //set max length for input Level 3
                            textViewInput.setFilters(new InputFilter[]{ new InputFilter.LengthFilter(5) });
                            leveltextview.setText("III");
                            renewWordLevel3();
                        }

                        else if(score >= 200)
                        {
                            //set max length for input Level 2
                            textViewInput.setFilters(new InputFilter[]{ new InputFilter.LengthFilter(4) });
                            leveltextview.setText("II" );
                            renewWordLevel2();
                        }
                        else
                        {
                            leveltextview.setText("I");
                            renewWord();
                        }

                        score += 10;//add the score
                        scoreTextView.setText("Score: " + score);//increase the score when user insert a right answer
                        plusTime();
                        //play sound
                        correctWord.start();

                    }
                    else
                    {
                     //added by Tammy Le
                        //When player gets a wrong word -> input will be vibrated
                        //Inspired by Priya
                        Animation vibrate = AnimationUtils.loadAnimation(getApplicationContext(), R.anim.vibrate);
                        textViewInput.startAnimation(vibrate);
                        //play sound incorrect
                        incorrectWord.start();
                    }

                }
                return false;
            }
        });
        /*
        initialize the click listeners for start button
        sound is played
        setTimer() function is called
        set the visibility of all the needed views
        start the timer by calling startTimer()
        */

        buttonStartTime.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                //play sound
                newgame.start();

                setTimer();
                score = 0;
                scoreTextView.setText("Score: " + score);
                leveltextview.setText("I");
                buttonStartTime.setVisibility(View.INVISIBLE);
                mProgressBar.setVisibility(View.INVISIBLE);
                textViewShuffle.setVisibility(View.VISIBLE);//show the word that user needs to guess
                scoreTextView.setVisibility(View.VISIBLE);// show the score when the game starts
                leveltextview.setVisibility(View.VISIBLE);// show the level when the game starts
                timerCoin.setVisibility(View.VISIBLE);//show the coin when the game starts
                buttonBuyTime.setVisibility(View.VISIBLE);// show the buyTime button when the game starts
                textViewInput.setVisibility(View.VISIBLE);//show the textInput


                startTimer();

                //set the visibility of the main progress bar
                mProgressBar1.setVisibility(View.VISIBLE);



            }
        });


        ///?Buy time button
        /*
        set the function to buy time when user click buytime Button
        there will be a condition
        if user's coin is more than 100 then they can buy more time buy trigger the plusTimer() function
        new amount of coin is also put into the SharedPreference to assure the consitency of coin through out all activities
         */
        buttonBuyTime.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(coin >= 100)
                {
                    coin = coin - 100;
                    editCoin(coin);// update the coin to the shared preference
                    plusTime();
                    timerCoin.setText("" + coin);
                }
                else
                {
                    Toast.makeText(getApplicationContext(), "You don't have enough coin to buy time", Toast.LENGTH_SHORT).show();
                }
            }
        });

        /*
        set click listener for highScore icon
        sound gets played and new intent is created to take user to TimerHighScorepActivity
         */
        highScoreTimer.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //play sound
                newgame.start();
                Intent intent = new Intent(getApplicationContext(), TimerHighScoreActivity.class);
                startActivity(intent);
                finish();
            }
        });

        /**
        *   set a click listener when user click on the coin or amount of coin left
        *   sound gets played and new intent is created to take user to ShopActivity
        **/
        timerShop.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //play sound
                newgame.start();
                Intent intent = new Intent(getApplicationContext(), ShopActivity.class);
                startActivity(intent);
                finish();
            }
        });

    }

    /*
    renewWord()
    get a new random word and shuffle all the letters to display
    parameters: no parameters needed
    return: void
    */
    private void renewWord() {
        givenWord = availableWords.randomWordLevel1();
        shuffedWord = availableWords.shuffleWord(givenWord);
        textViewShuffle.setText(shuffedWord);
    }
    //different renewWord function to change to more complex word when user hitting a curtain score
        /*
    renewWordLevel2()
    get a new random word from WORDLEVEL2 array in LetterStorage class and shuffle all the letters to display
    parameters: no parameters needed
    return: void
     */
    private void renewWordLevel2()
    {
        givenWord = availableWords.randomWordLevel2();
        shuffedWord = availableWords.shuffleWord(givenWord);
        textViewShuffle.setText(shuffedWord);
    }
    /*
   renewWordLevel3()
  get a new random word from WORDLEVEL3 array in LetterStorage class and shuffle all the letters to display
  parameters: no parameters needed
  return: void
 */
    private void renewWordLevel3()
    {
        givenWord = availableWords.randomWordLevel3();
        shuffedWord = availableWords.shuffleWord(givenWord);
        textViewShuffle.setText(shuffedWord);
    }

    //Added by Tammy Le, 13/5/2019
            /*
    renewWordLevel4()
    get a new random word from WORDLEVEL4 array in LetterStorage class and shuffle all the letters to display
    parameters: no parameters needed
    return: void
     */
    private void renewWordLevel4()
    {
        givenWord = availableWords.randomWordLevel4();
        shuffedWord = availableWords.shuffleWord(givenWord);
        textViewShuffle.setText(shuffedWord);
    }

    /**
     *  function setTimer: is used to initialize values for the timer
     *  parameters: no parameters needed for this funtion
     *  returns : void
     *  const: time --> set the time to 10 which will be multiplied by 1000 to make the timer to 10 seconds*/
    private void setTimer(){

        //initialize the time variable
        int time = 10;

        //calculate the totaltime
        totalTimeCountInMilliseconds =  time * 1000;

        //set the maximum time of the progress bar timer
        mProgressBar1.setMax( time * 1000);
    }

    /**
     * function startTimer: is used to start counting down the timer
     * parameters: no parameters needed for this function
     * returns : void
     * */

    private void startTimer() {

        //call renewWord function once the timer is on
        renewWord();
        countDownTimer = new MyCountDown(totalTimeCountInMilliseconds, 50);
        countDownTimer.start();
    }

    /**
     *   function plusTime: is used to give extra time whenever player gets correct word
     *   parameters: no parameters needed for this function
     *   returns : void
     */
    private void plusTime(){

        if(totalTimeCountInMilliseconds < 10000)
        {
            countDownTimer.cancel();
            countDownTimer = new MyCountDown(totalTimeCountInMilliseconds + 3000, 50);// give extra 3000 milliseconds if they get the answer right
        }
        else
        {
            countDownTimer.cancel();
            countDownTimer = new MyCountDown(10000, 50);// set the timer to 10000 millisec if totalTimeCountInMilliseconds + 3000 > 10000
        }

        countDownTimer.start();


    }
//function to close the soft keyboard
    /*
    closeKeyboard: use to close the softkeyboard when its called
    parameters: no parameters needed
    returns: void, the function just closes the keyboard, no return value needed
     */
    private void closeKeyboard()
    {
        View view = this.getCurrentFocus();
        if(view != null)
        {
            InputMethodManager imm = (InputMethodManager)getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }
    //create a custom CountDownTimer
    /*
    a new class created extended from CountDownTimer
    we customed the onTick and onFinish functions to be suitable for our own app purpose
    like set visibility of some elememts for them to appear in appropriate way
    or set the progress bar along with the timer
     */
    public class MyCountDown extends CountDownTimer {
        public MyCountDown(long millisInFuture, long countDownInterval) {
            super(millisInFuture, countDownInterval);
        }

        @Override
        public void onTick(long l) {
            totalTimeCountInMilliseconds = l;
            mProgressBar1.setProgress((int) totalTimeCountInMilliseconds);
        }

        @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
        @Override
        public void onFinish() {
            //set up the visibility of all needed views
            textViewInput.setVisibility(View.INVISIBLE);
            buttonStartTime.setVisibility(View.VISIBLE);
            mProgressBar.setVisibility(View.VISIBLE);
            mProgressBar1.setVisibility(View.GONE);
            textViewInput.setText("");
            textViewShuffle.setText(givenWord);
            buttonBuyTime.setVisibility(View.INVISIBLE);

            //put the new score into highScore list (if suitable)

            SharedPreferences preferences = getSharedPreferences("TIME_PREFS", 0);
            SharedPreferences.Editor editor = preferences.edit();
            editor.putInt("latestscore", score);
            editor.apply();

            //Hide the soft keyboard when timer finishes
//            textViewInput.onEditorAction(EditorInfo.IME_ACTION_DONE);

            closeKeyboard();

        }


    }
    /*
    editCoin()
    take the current amount of coin and put it into the SharedPreferences
    parameters: current number of coin in Integer form
    return: void, the function just edit the coin in the SharedPreference, doesn't return anything
    */
    public void editCoin(int coin)
    {
        SharedPreferences preferences = getSharedPreferences("COIN_PREFS", 0);
        SharedPreferences.Editor editor = preferences.edit();
        editor.putInt("coin", coin);
        editor.apply();
    }


    /*
    retrieveCoin()
    retrieve the amount of coin currently holding in the SharedPreference
    parameters: no parameters needed
    returns: integer contains the number of coins.
     */
    public int retrieveCoin()
    {
        int coin1;
        SharedPreferences preferences = getSharedPreferences("COIN_PREFS", 0);

        coin1 = preferences.getInt("coin", 0);
        return coin1;
    }


}